runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/FnApiControlClient.java

```diff
@@ -51,23 +54,25 @@
   private final ResponseStreamObserver responseObserver = new ResponseStreamObserver();
   private final ConcurrentMap<String, CompletableFuture<BeamFnApi.InstructionResponse>>
       outstandingRequests;
+  private final Set<Consumer<FnApiControlClient>> onCloseListeners = ConcurrentHashMap.newKeySet();
```

Comment:
This seems like it'll do the right thing, but gosh it's really unfortunate that Java doesn't have a cleaner way to create a Concurrent Set

Reply:
Yup, even ConcurrentHashMap.newKeySet() is a new addition to java8.